pkg/security/san_pinning.go

```diff
@@ -0,0 +1,172 @@
+package security
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1alpha1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/runtime-watcher/listener/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	debugLogLevel    = 2
+	requestSizeLimit = 16000
+
+	xfccHeader           = "X-Forwarded-Client-Cert"
+	headerValueSeparator = ";"
+	keyValueSeparator    = "="
+	certificateKey       = "Cert="
+
+	shootDomainKey = "skr-domain"
+)
+
+var (
+	// Static errors.
+	errNotVerified   = errors.New("SAN from certificate does not match domain specified in KymaCR")
+	errPemDecode     = errors.New("failed to decode PEM block")
+	errEmptyCert     = errors.New("empty certificate")
+	errHeaderMissing = fmt.Errorf("request does not contain '%s' header", xfccHeader)
+)
+
+type RequestVerifier struct {
+	Client client.Client
+	log    logr.Logger
+}
+
+func NewRequestVerifier(client client.Client) *RequestVerifier {
+	return &RequestVerifier{
+		Client: client,
+		log:    ctrl.Log.WithName("request–verifier"),
+	}
+}
+
+// Verify verifies the given request by fetching the KymaCR given in the request payload
+// and comparing the SAN(subject alternative name) of the certificate with the SKR-domain of the KymaCR.
+// If the request can be verified 'nil' will be returned.
+func (v *RequestVerifier) Verify(request *http.Request) error {
+	certificate, err := v.getCertificateFromHeader(request)
+	if err != nil {
+		return err
+	}
+
+	domain, err := v.getDomain(request)
+	if err != nil {
+		return err
+	}
+
+	if contains(certificate.URIs, domain) ||
```

Comment:
This verification logic should be extracted to a separate function and unit-tested to properly document what is a "matching" combination and what is not.
For example, does the client using certificate issued for ```client.example.domain.com``` match the Kyma configured for ```example.domain.com```?